src/mpi/init/initthread.c

```diff
@@ -669,6 +669,14 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
 #if defined(MPICH_IS_THREADED)
     if (cs_initialized) {
         MPIR_Thread_CS_Finalize();
+
+        /* Finalize the threading library after releasing all synchronization
+         * objects (e.g., mutexes) */
+        {
+            int thread_err;
+            MPL_thread_finalize(&thread_err);
```

Comment:
@shintaro-iwasaki I did move this call to inside an `MPICH_IS_THREADED` conditional here.

Reply:
Oh, sorry. I didn't realize this block was actually inside an init function... Yes, we will need this in `MPI_Finalize`. Can you draft a patch to do that?

Reply:
Sure!

Reply:
Please `git apply` the following diff:
```
diff --git a/src/mpi/init/finalize.c b/src/mpi/init/finalize.c
index 1180940..0fb257c 100644
--- a/src/mpi/init/finalize.c
+++ b/src/mpi/init/finalize.c
@@ -331,6 +331,16 @@ int MPI_Finalize(void)
     }
 #endif
 
+#if defined(MPICH_IS_THREADED)
+    /* Finalize the threading library after releasing all synchronization
+     * objects (e.g., mutexes) */
+    {
+        int thread_err;
+        MPL_thread_finalize(&thread_err);
+        MPIR_Assert(thread_err == 0);
+    }
+#endif
+
     /* ... end of body of routine ... */
   fn_exit:
     MPIR_FUNC_TERSE_FINALIZE_EXIT(MPID_STATE_MPI_FINALIZE);
diff --git a/src/mpi/init/initthread.c b/src/mpi/init/initthread.c
index 4011aee..a966417 100644
--- a/src/mpi/init/initthread.c
+++ b/src/mpi/init/initthread.c
@@ -669,14 +669,6 @@ int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
 #if defined(MPICH_IS_THREADED)
     if (cs_initialized) {
         MPIR_Thread_CS_Finalize();
-
-        /* Finalize the threading library after releasing all synchronization
-         * objects (e.g., mutexes) */
-        {
-            int thread_err;
-            MPL_thread_finalize(&thread_err);
-            MPIR_Assert(thread_err == 0);
-        }
     }
 #endif
     return mpi_errno;
```